### PR TITLE
Base Fee Floor (Fixed)

### PR DESCRIPTION
**Pull request by gastonponti**
**Description**

Adds a way to set a fixed BaseFeeFloor. If the baseFee goes below that parameter, return it instead

**Tests**

Added tests for the edge case

**Additional context**

Also changed a check of the Cel2 activation that could had some issues if the time was not exactly the same

**Metadata**

- Related issue: https://github.com/celo-org/celo-blockchain-planning/issues/478

### CODE REVIEW

**Comment by palango** [comment]
Two questions which come to mind:
- Do we need changes in the txpool to reject (or somehow handle) transactions with too low fees)?
- How would this be set in practice? Do we hardcode the config?

File: consensus/misc/eip1559/eip1559.go [hunk 1]
```diff
@@ -60,7 +60,11 @@ func VerifyEIP1559Header(config *params.ChainConfig, parent, header *types.Heade
 func CalcBaseFee(config *params.ChainConfig, parent *types.Header, time uint64) *big.Int {
 	// If this is the cel2 transition block and the parent block has a base fee
 	// then use that.
-	if config.Cel2Time != nil && *config.Cel2Time == time && parent.BaseFee != nil {
+	if !config.IsCel2(parent.Time) && config.IsCel2(time) && parent.BaseFee != nil {
```
Comment (palango) [inline consensus/misc/eip1559/eip1559.go]: I still think this wouldn't be triggered on any migrated chain. In the migration we set the Cel2 flag on the migration block, so every parent will have it enabled: https://github.com/celo-org/optimism/blob/celo8/op-chain-ops/cmd/celo-migrate/state.go#L254

Besides that, do we need the change here? I think the change in `CalcBaseFee` should be enough.

Comment (gastonponti) [reply consensus/misc/eip1559/eip1559.go]: So that if is never going to be reached?

What I'm doing with that if, it's required to "normalize" the baseFee, if for some reason we have it under the floor, if the baseFee increases or it's the same, will be under the floor until one block tries to reduce the baseFee.

That would happen only if the baseFee is under the Floor, which shouldn't happen because we are going to set it anyways and we should set the same Floor that the existing network has. So, for example, if the new floor is 5, and the baseFee is in 3 in the cel2 switch, without that correction,  if it keeps using the targetDensity amount of gas, will be in 3 instead of 5 as long as it hits less than the target density.

The other possibility, is to add a check in the equals and when increases, but I think that we can avoid those checks if we correct it in the block switch (and again, the only possible scenario is if we set a Floor smaller than the Floor that the testnet has)

**Comment by gastonponti** [comment]
> * Do we need changes in the txpool to reject (or somehow handle) transactions with too low fees)?

We could, but I'm not seeing it necessary right now. Eventually we will add the dynamic usd floor, and we can think to add that. Adding that check it also requires to add some currency conversion in the tx pool. Right now every node will eventually reject txs that have more than X time, or prune it with new txs with higher baseFee.
I tried to stick to the minimum changes required for Alfajores. That check could be added later without a fork.

> * How would this be set in practice? Do we hardcode the config?

I'm going to set it with optimism. I tried to use the same structure as the TargetDensity changes. If we add that as a parameter, missing that same config will stall that node, so I prefer to add it as part of the "hard configuration" fields

File: consensus/misc/eip1559/eip1559.go (hunk 1 above)
Comment (mcortesi) [reply consensus/misc/eip1559/eip1559.go]: i don't get why we are checking 'time' here. shouldn't we just check a "feature flag" here?

like `if config.baseFeeFloor >0` then...
a call it a day?

and then add a `Config.BaseFeeFloor` option

**Comment by mcortesi** [comment]
Similar to @palango , it's unclear to me how the `ChainConfig` is configure. I believe it actually comes from the genesis.json. Which means we need to add it there.

If so, we need to make sure that whereever we generate genesis.json we also include this.

**Comment by gastonponti** [comment]
Related to how we can create that configuration, I've created this PR in celo-org/optimism
https://github.com/celo-org/optimism/pull/231

File: consensus/misc/eip1559/eip1559.go (hunk 1 above)
Comment (gastonponti) [reply consensus/misc/eip1559/eip1559.go]: Refactored the code to avoid any modification on what we already had, and just check the return value against a feature flag

File: params/config.go [hunk 2]
```diff
@@ -491,6 +493,16 @@ func (o *OptimismConfig) String() string {
 	return "optimism"
 }
 
+// CeloConfig is the celo config.
```
Comment (karlb) [inline params/config.go]: This doc string is trivial and should be improved or omitted.

File: params/config.go [hunk 3]
```diff
@@ -491,6 +493,16 @@ func (o *OptimismConfig) String() string {
...
+type CeloConfig struct {
+	EIP1559BaseFeeFloor uint64 `json:"eip1559BaseFeeFloor"`
```
Comment (karlb) [inline params/config.go:502]: Is it only me, or does putting EIP-1559 into the name make is sound like it only applies to EIP-1559 txs and not legacy txs?

File: consensus/misc/eip1559/eip1559.go [hunk 4]
```diff
@@ -57,7 +57,14 @@ func VerifyEIP1559Header(config *params.ChainConfig, parent, header *types.Heade
...
 // CalcBaseFee calculates the basefee of the header.
 // The time belongs to the new block to check if Canyon is activted or not
-func CalcBaseFee(config *params.ChainConfig, parent *types.Header, time uint64) *big.Int {
+func CalcBaseFee(config *params.ChainConfig, parent *types.Header, time uint64) (response *big.Int) {
+	defer func() {
+		// If the base fee response is below the floor, intercept the return and return the floor instead.
+		if config.Celo != nil {
+			response = math.BigMax(response, new(big.Int).SetUint64(config.Celo.EIP1559BaseFeeFloor))
+		}
+	}()
```
Comment (karlb) [inline consensus/misc/eip1559/eip1559.go:67]: Using a defer is a creative way to handle all the different returns without refactoring the function. A bit unintuitive but clever!

**Review by karlb — APPROVED** [review]

File: params/config.go (hunk 3 above)
Comment (gastonponti) [reply params/config.go:502]: I've tried to maintain the same way that optimism is naming the base fee parameters. The eip1559 is about the baseFee, maybe the Tx shouldn't have the 1559 on it, haha.

File: consensus/misc/eip1559/eip1559.go (hunk 4 above)
Comment (gastonponti) [reply consensus/misc/eip1559/eip1559.go:67]: I agree that is not intuitive. I will add another comment more as a warning for the return, just in case

File: params/config.go [hunk 5]
```diff
@@ -491,6 +493,15 @@ func (o *OptimismConfig) String() string {
 	return "optimism"
 }
 
+type CeloConfig struct {
+	EIP1559BaseFeeFloor uint64 `json:"eip1559BaseFeeFloor"`
+}
+
+// String implements the stringer interface, returning the celo config details.
+func (o *CeloConfig) String() string {
+	return "celo"
```
Comment (carterqw2) [inline params/config.go]: Is it some sort of a weird convention to just return this instead of actual details?

Comment (karlb) [reply params/config.go]: There is plenty of precedent in the same file: https://github.com/celo-org/op-geth/blob/715ccb804815f2cbee46c773ec4cbe2f32b50b2d/params/config.go#L466-L492

But it is a bit strange.

Comment (gastonponti) [reply params/config.go]: Yes, I just mimicked from the Optimism one, but maybe they also dragged a bad convention. I'll add it (for us at least)
